parquet-variant/src/builder.rs

```diff
@@ -835,18 +936,23 @@ impl<'a> ObjectBuilder<'a> {
     ///
     /// Note: when inserting duplicate keys, the new value overwrites the previous mapping,
     /// but the old value remains in the buffer, resulting in a larger variant
-    pub fn insert<'m, 'd, T: Into<Variant<'m, 'd>>>(&mut self, key: &str, value: T) {
+    pub fn insert<'m, 'd, T: Into<Variant<'m, 'd>>>(
+        &mut self,
+        key: &str,
+        value: T,
+    ) -> Result<(), ArrowError> {
         // Get metadata_builder from parent state
         let metadata_builder = self.parent_state.metadata_builder();
 
-        let field_id = metadata_builder.upsert_field_name(key);
+        let field_id = metadata_builder.upsert_field_name(key)?;
         let field_start = self.buffer.offset();
 
         if self.fields.insert(field_id, field_start).is_some() && self.validate_unique_fields {
             self.duplicate_fields.insert(field_id);
         }
```

Comment:
Now that `insert` is fallible, we can fail-fast instead of tracking a set of duplicate fields for `finish` to worry about:
```suggestion
        if self.fields.insert(field_id, field_start).is_some() && self.validate_unique_fields {
            return Err(... duplicate field {key} ...);
        }
```